runtime/service_integ_test.go

```diff
@@ -1463,6 +1466,73 @@ func TestStopVM_Isolated(t *testing.T) {
 		})
 	}
 }
+func TestEvents_Isolated(t *testing.T) {
+	prepareIntegTest(t)
+	require := require.New(t)
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	// If we don't have enough events within 30 seconds, the context will be cancelled and the loop below will be interrupted
+	subscribeCtx, subscribeCancel := context.WithTimeout(ctx, 30*time.Second)
+	defer subscribeCancel()
+	eventCh, errCh := client.Subscribe(subscribeCtx, "topic")
+
+	image, err := alpineImage(ctx, client, defaultSnapshotterName())
+	require.NoError(err, "failed to get alpine image")
+
+	pluginClient, err := ttrpcutil.NewClient(containerdSockPath + ".ttrpc")
+	require.NoError(err, "failed to create ttrpc client")
+
+	vmID := testNameToVMID(t.Name())
+
+	fcClient := fccontrol.NewFirecrackerClient(pluginClient.Client())
+	_, err = fcClient.CreateVM(ctx, &proto.CreateVMRequest{VMID: vmID})
+	require.NoError(err)
+
+	c, err := client.NewContainer(ctx,
+		"container-"+vmID,
+		containerd.WithSnapshotter(defaultSnapshotterName()),
+		containerd.WithNewSnapshot("snapshot-"+vmID, image),
+		containerd.WithNewSpec(oci.WithProcessArgs("/bin/echo", "-n", "hello"), firecrackeroci.WithVMID(vmID)),
+	)
+	require.NoError(err)
+
+	stdout := startAndWaitTask(ctx, t, c)
+	require.Equal("hello", stdout)
+
+	_, err = fcClient.StopVM(ctx, &proto.StopVMRequest{VMID: vmID})
+	require.Equal(status.Code(err), codes.OK)
+
+	expected := []string{
+		"/snapshot/prepare",
+		"/snapshot/commit",
+		"/firecracker-vm/start",
+		"/snapshot/prepare",
+		"/containers/create",
+		"/tasks/create",
+		"/tasks/start",
+		"/tasks/exit",
+		"/tasks/delete",
+		"/firecracker-vm/stop",
+	}
+	var actual []string
+
+loop:
+	for len(actual) < len(expected) {
+		select {
+		case event := <-eventCh:
+			actual = append(actual, event.Topic)
+		case err := <-errCh:
+			assert.NoError(t, err)
+			break loop
```

Comment:
Will this correctly exit in the event that actual never reaches the appropriate size? It seems like it will get stuck in the select statement potentially waiting for an event that might not come.

Reply:
The Subscribe() call above is having a context with deadline. The gorountine inside containerd will be cancelled if we don't have enough events within 30 seconds.

https://github.com/containerd/containerd/blob/c20c569a01a524a3b808848e210383d7cd60ef9c/events.go#L77

Reply:
Thanks! I missed the context timeout.